inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/management/DynamicWorkloadManagementPage.java

```diff
@@ -384,4 +618,92 @@ private void actionReset(AjaxRequestTarget aTarget, Form<?> aForm)
 
         aTarget.add(searchForm);
     }
+
+    private void actionConfirm(AjaxRequestTarget aTarget, Form<?> aForm)
+    {
+        aTarget.addChildren(getPage(), IFeedback.class);
+        dynamicWorkloadExtension.writeTraits(workloadManagementService,
+                new WorkloadTraits(workflowChoices.getDefaultModelObjectAsString(),
+                        Integer.parseInt(defaultNumberDocumentsTextField.getInput())),
+            currentProject.getObject());
+        success("Changes saved");
+    }
+
+    private void actionSetDocumentStatus(AjaxRequestTarget aAjaxRequestTarget,
+            Form<AnnotationDocumentState> aForm)
+    {
+        aAjaxRequestTarget.addChildren(getPage(), IFeedback.class);
+        if (resetDocument.getModelObject() == null) {
+            error("Please select a document you wish to change its status first!");
+        }
+        else if (documentState.getModelObject() == null) {
+            error("Please select a state you wish to assign to the document");
+        }
+        else {
+            if (documentState.getModelObject().equals(AnnotationDocumentState.IN_PROGRESS)) {
+                documentService.transitionAnnotationDocumentState(
+                        documentService.getAnnotationDocument(
+                                documentService.getSourceDocument(
+                                        currentProject.getObject(),
+                                        resetDocument.getModelObject().getName()),
+                                userSelection.getModelObject()),
+                        AnnotationDocumentStateTransition.
+                            ANNOTATION_FINISHED_TO_ANNOTATION_IN_PROGRESS);
+            }
+            else {
+                documentService.transitionAnnotationDocumentState(
+                        documentService.getAnnotationDocument(
+                                documentService.getSourceDocument(
+                                        currentProject.getObject(),
+                                        resetDocument.getModelObject().getName()),
+                                userSelection.getModelObject()),
+                        AnnotationDocumentStateTransition.
+                            ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
+            }
+
+            success("Document status changed");
+        }
+    }
+
+    private void actionAssignDocument(AjaxRequestTarget aAjaxRequestTarget,
+            Form<List<SourceDocument>> aForm)
+    {
+        aAjaxRequestTarget.addChildren(getPage(), IFeedback.class);
+
+        if (aForm.getModelObject() == null) {
+            error("Please add documents you want to assign.");
+        }
+        else {
+
+            for (SourceDocument sourceDocument : aForm.getModelObject()) {
```

Comment:
The body of this loop looks very much like backend code that should go into a service bean...

Reply:
I implemented it the same way as it was done in webanno (e.g. UserSelectionPanel). But I can try to change that ofc

Reply:
I guess the `UserSelectionPanel` is probably one of the oldest pieces of code we have - I don't know how often it was refactored and in particular with respect to performance/scalability. The workflow changes we make here have the goal of supporting a scalable workflow with lots of users and lots of documents, so we need to keep in mind that DB accesses e.g. over source/annotation documents need to be efficient as well.

Reply:
Hehe ok, sure I'll change that